src/Framework/Hooks/CommonConditions.cs

```diff
@@ -123,67 +126,145 @@ private static QuestStatSummary GetQuestStats(CustomQuest managedQuest)
-
-                flag &= currentLevel >= expectedLevel;
-                
-                if (Monitor.IsVerbose)
-                    Monitor.Log(
-                        $"Checked friendship level for `{who}`, " +
-                        $"current level: {currentLevel}, " +
-                        $"expected: {expectedLevel}, " +
-                        $"current flag: {flag}");
-            }
-
-            return flag;
-        }
-
+        public static bool CheckEventSeenCondition(string valueToCheck)
+        {
+            int[] events = Utility.parseStringToIntArray(valueToCheck);
+            bool flag = true;
+
+            if (events.Length < 1)
+                return false;
+
+            foreach (var ev in events)
+            {
+                flag &= Game1.player.eventsSeen.Contains(ev);
+                Monitor.VerboseLog($"Checked if event `{ev}` was seen. Current flag: {flag}");
+            }
+
+            return flag;
+        }
+
+        public static bool CheckReceivedMailCondition(string valueToCheck)
+        {
+            string[] mails = valueToCheck.Split(' ');
+            bool flag = true;
+
+            if (mails.Length < 1)
+                return false;
+
+            foreach (string mail in mails)
+            {
+                flag &= Game1.player.mailReceived.Contains(mail);
+                Monitor.VerboseLog($"Checked if mail letter `{mail}` was received. Current flag: {flag}");
+            }
+
+            return flag;
+        }
+
+        public static bool CheckFriendshipLevel(string friendshipLevel)
+        {
+            string[] flevel = friendshipLevel.Split(' ');
+            bool flag = true;
+
+            if (flevel.Length < 2)
+                return false;
+
+            for (int i = 0; i < flevel.Length; i += 2)
+            {
+                string whoLevel = flevel[i];
+                int currentFriendshipLevel = Game1.player.getFriendshipHeartLevelForNPC(whoLevel);
+                int expectedFriendshipLevel = Convert.ToInt32(flevel[i + 1]);
+
+                flag &= currentFriendshipLevel >= expectedFriendshipLevel;
+
+                if (Monitor.IsVerbose)
+                    Monitor.Log(
+                        $"Checked friendship level for `{whoLevel}`, " +
+                        $"current level: {currentFriendshipLevel}, " +
+                        $"expected: {expectedFriendshipLevel}, " +
+                        $"current flag: {flag}");
+            }
+
+            return flag;
+        }
+
+        public static bool CheckFriendshipStatus(string friendshipStatus) 
+        {
+            string[] fstatus = friendshipStatus.Split(' ');
+            bool flag = true;
+
+            if (fstatus.Length < 2)
+                return false;
+
+            for (int i = 0; i < fstatus.Length; i += 2)
+            {
+                string whoStatus = fstatus[i];
+                string currentStatus = Game1.player.friendshipData[whoStatus].Status.ToString();

```

Comment:
Please add check for key exist, for the `Game1.player.friendshipData[whoStatus]`, like

```cs
string whoStatus = fstatus[i];

if (!Game1.player.friendshipData.ContainsKey(whoStatus))
{
  flag &= false;
  continue;
}

string currentStatus = Game1.player.friendshipData[whoStatus].Status.ToString();
// ...
```

Otherwise this line throws an exception, when key not exists in dictionary.

Reply:
I think the line alr added and when I tested using random name and the red error line doesn't show up (is those called exception?)